File: swift/Sources/TrailCore/Playback.swift
import Foundation

public enum TrailPlaybackStatus: Sendable { case playing, paused, finished }
/// Value-type deterministic player. UI ownership and frame scheduling belong to platform adapters.
public struct TrailPlayer: Sendable {
    public private(set) var effect: TrailEffect
    public private(set) var elapsedSeconds = 0.0
    public private(set) var status: TrailPlaybackStatus
    private var seekEndpoint = false
    public init(effect: TrailEffect = TrailEffect(), autoPlay: Bool = true) {
        self.effect = effect; self.status = autoPlay && effect.durationSeconds > 0 ? .playing : .paused
    }
    public var progress: Double {
        guard effect.durationSeconds > 0 else { return 1 }
        let cycles = elapsedSeconds / effect.durationSeconds
        return effect.repeats && !seekEndpoint ? cycles - floor(cycles) : min(1, max(0, cycles))
    }
    public mutating func play() {
        guard effect.durationSeconds > 0 else { return }
        seekEndpoint = false; if status == .finished { elapsedSeconds = 0 }; status = .playing
    }
    public mutating func pause() { if status == .playing { status = .paused } }
    public mutating func replay() { elapsedSeconds = 0; seekEndpoint = false; status = effect.durationSeconds > 0 ? .playing : .paused }
    public mutating func seek(to fraction: Double) { fractionCheck(fraction); seekEndpoint = fraction == 1; elapsedSeconds = fraction * effect.durationSeconds; status = .paused }
    public mutating func advance(by seconds: Double) {
        precondition(seconds.isFinite && seconds >= 0)
        guard status == .playing else { return }
        let next = elapsedSeconds + seconds; precondition(next.isFinite)
        elapsedSeconds = effect.repeats ? next : min(next, effect.durationSeconds)
        if !effect.repeats && elapsedSeconds >= effect.durationSeconds { status = .finished }
    }
    public mutating func configure(_ effect: TrailEffect, reset: Bool = false) {
        let normalized = self.effect.durationSeconds == 0 ? 0 : elapsedSeconds / self.effect.durationSeconds
        self.effect = effect
        if reset { seekEndpoint = false }
        elapsedSeconds = reset ? 0 : normalized * effect.durationSeconds
        if effect.durationSeconds == 0 { status = .paused } else if reset { status = .playing }
    }
    public func frame(layer: Int, reducedMotion: Bool = false) -> TrailVisualState {
        if !reducedMotion, seekEndpoint, let spec = effect.layers[layer].animation, elapsedSeconds == spec.durationSeconds {
            return spec.sampler.sample(at: TrailTime(progress: 1, elapsedSeconds: spec.durationSeconds))
        }
        return effect.sample(layer: layer, elapsed: elapsedSeconds, reducedMotion: reducedMotion)
    }
}

File: android/trail-core/src/main/kotlin/dev/trail/core/Playback.kt
package dev.trail.core

enum class TrailPlaybackStatus { Playing, Paused, Finished }

/** Host-owned deterministic clock. UI adapters drive it only while visible and active. */
class TrailPlayer(effect: TrailEffect = TrailEffect(), autoPlay: Boolean = true) {
    var effect: TrailEffect = effect; private set
    var elapsedSeconds: Double = 0.0; private set
    private var seekEndpoint = false
    var status: TrailPlaybackStatus = if (autoPlay && effect.durationSeconds > 0) TrailPlaybackStatus.Playing else TrailPlaybackStatus.Paused; private set
    val progress: Double get() = if (effect.durationSeconds == 0.0) 1.0 else {
        val cycles = elapsedSeconds / effect.durationSeconds
        if (effect.repeats && !seekEndpoint) cycles - kotlin.math.floor(cycles) else cycles.coerceIn(0.0, 1.0)
    }
    fun play() { if (effect.durationSeconds == 0.0) return; seekEndpoint = false; if (status == TrailPlaybackStatus.Finished) elapsedSeconds = 0.0; status = TrailPlaybackStatus.Playing }
    fun pause() { if (status == TrailPlaybackStatus.Playing) status = TrailPlaybackStatus.Paused }
    fun replay() { elapsedSeconds = 0.0; seekEndpoint = false; status = if (effect.durationSeconds > 0) TrailPlaybackStatus.Playing else TrailPlaybackStatus.Paused }
    fun seek(fraction: Double) { requireFraction(fraction, "seek"); seekEndpoint = fraction == 1.0; elapsedSeconds = fraction * effect.durationSeconds; status = TrailPlaybackStatus.Paused }
    fun advance(seconds: Double) {
        require(seconds.isFinite() && seconds >= 0) { "Clock delta must be finite and nonnegative" }
        if (status != TrailPlaybackStatus.Playing) return
        val next = elapsedSeconds + seconds
        require(next.isFinite()) { "Clock overflow" }
        elapsedSeconds = if (effect.repeats) next else next.coerceAtMost(effect.durationSeconds)
        if (!effect.repeats && elapsedSeconds >= effect.durationSeconds) status = TrailPlaybackStatus.Finished
    }
    fun configure(effect: TrailEffect, reset: Boolean = false) {
        val oldDuration = this.effect.durationSeconds
        val normalized = if (oldDuration == 0.0) 0.0 else elapsedSeconds / oldDuration
        this.effect = effect
        if (reset) seekEndpoint = false
        elapsedSeconds = if (reset) 0.0 else normalized * effect.durationSeconds
        if (effect.durationSeconds == 0.0) status = TrailPlaybackStatus.Paused
        else if (reset) status = TrailPlaybackStatus.Playing
    }
    fun frame(layer: Int, reducedMotion: Boolean = false): TrailVisualState {
        // Paused seek-to-end displays the final frame, even for a looping clip.
        val spec = effect.layers[layer].animation
        if (!reducedMotion && seekEndpoint && spec != null && elapsedSeconds == spec.durationSeconds) {
            return spec.sampler.sample(TrailTime(1.0, 0, spec.durationSeconds))
        }
        return effect.sample(layer, elapsedSeconds, reducedMotion)
    }
}
